libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -2351,6 +2351,44 @@ string YulUtilFunctions::copyArrayFromStorageToMemoryFunction(ArrayType const& _
 	});
 }
 
+string YulUtilFunctions::bytesConcatFunction(vector<Type const*> const& _argumentTypes)
+{
+	string functionName = "bytes_concat";
+	size_t totalParams = 0;
+	for (Type const* argumentType: _argumentTypes)
+	{
+		solAssert(
+			argumentType->isImplicitlyConvertibleTo(*TypeProvider::bytesMemory()) ||
+			argumentType->isImplicitlyConvertibleTo(*TypeProvider::fixedBytes(32)),
+			""
+		);
+		totalParams += argumentType->sizeOnStack();
+		functionName += "_" + argumentType->identifier();
+	}
+
+	return m_functionCollector.createFunction(functionName, [&]() {
+		Whiskers templ(R"(
+			function <functionName>(<parameters>) -> outPtr {
+				outPtr := <allocateUnbounded>()
+				let dataStart := add(outPtr, 0x20)
+				let dataEnd := <encodePacked>(dataStart<?+parameters>, <parameters></+parameters>)
+				mstore(outPtr, sub(dataEnd, dataStart))
+				<finalizeAllocation>(outPtr, sub(dataEnd, outPtr))
+			}
+		)");
+		templ("functionName", functionName);
+		string paramPrefix = "param_";
+		templ("parameters", suffixedVariableNameList(paramPrefix, 0, totalParams));
+		templ("allocateUnbounded", allocateUnboundedFunction());
+		templ("finalizeAllocation", finalizeAllocationFunction());
+		templ(
+			"encodePacked",
+			ABIFunctions{m_evmVersion, m_revertStrings, m_functionCollector}.tupleEncoderPacked(_argumentTypes, _argumentTypes)
```

Comment:
For `abi.encodePacked` we do stuff like calling `fullEncodingType`.
Does `tupleEncoderPacked` acutally work for a storage bytes type?
Shouldn't we be more careful here when selecting the target types? I was thinking about specifically selecting bytesMemory or one of the bytesNN types.

Reply:
I set actual argument type if it is `FixedBytes` type, and `bytes memory` in other cases
Also, adding test for storage ref